### PR TITLE
Rescale joystick axis values according to deadzones

### LINKED ISSUE #3712: Scale joystick axis values according to deadzones
## Current situation

Currently, the way axis deadzones work in the framework is:

- If the current axis value is within the deadzone, the provided axis value is 0
- Otherwise, the provided axis value is the raw value

In the situation where the deadzone is large (say, 0.5), reported axis values would be as follows:

|Raw value|Provided value after deadzone application|
|---|---|
|0|0|
|0.25|0|
|0.5|0.5|
|0.75|0.75|
|1|1|

image

## My suggestion

Scaling them down as such would mitigate the sudden jumps:

|Raw value|Provided value after deadzone and scale|
|---|---|
|0|0|
|0.25|0|
|0.5|0|
|0.75|0.5|
|1|1|

image

Here's an [interactive GeoGebra file](https://www.geogebra.org/graphing/qjuhtxnv) to help visualize the idea.

### PR DESCRIPTION
**Pull request by Jeremiidesu**
Closes #3712.

Only issue left seems to be rounding errors that are particularly visible on large deadzones, which is fairly easy to replicate using triggers.

On my Dualshock 4, extreme values on triggers (deadzone `0.999f`) without clamping are `-1.0152590` and `0.9847410`.

### CODE REVIEW

File: osu.Framework/Input/Handlers/Joystick/OsuTKJoystickHandler.cs [hunk 1]
```diff
@@ -221,7 +224,8 @@ public void UpdateRawState(osuTK.Input.JoystickState rawState)
                         if (Precision.AlmostEquals(0, axisValue))
                             continue;
 
-                        defaultDeadZones.Value[i] = axisValue + deadzone_threshold;
+                        // Cap deadzones at 0.999f to avoid division by zero when rescaling
+                        defaultDeadZones.Value[i] = Math.Min(0.999f, axisValue + deadzone_threshold);
```
Comment (bdach) [inline osu.Framework/Input/Handlers/Joystick/OsuTKJoystickHandler.cs:228]: With the aforementioned rounding anomalies (which are due to [catastrophic cancellation](https://en.wikipedia.org/wiki/Loss_of_significance) and which I can't see a way out of) when calculating `1 - deadzone`) I'd be in favour of capping this even lower (`0.95f`? `0.9f`? maybe even `0.5f`?)

Also if we're sanity-checking this, this should probably be clamped from the bottom (to 0, at least).

Comment (Jeremiidesu) [reply osu.Framework/Input/Handlers/Joystick/OsuTKJoystickHandler.cs:228]: I guess realistically deadzones shouldn't exceed `0.5f` so I'll go for that.

As for clamping from the bottom, I don't understand what you have in mind by that.

Comment (bdach) [reply osu.Framework/Input/Handlers/Joystick/OsuTKJoystickHandler.cs:228]: Well, the deadzone should be within the range of [0,1], right? I'd feel safer with this change if that was explicitly clamped from both sides here (although not sure whether that's a plausible scenario to happen).

Comment (Jeremiidesu) [reply osu.Framework/Input/Handlers/Joystick/OsuTKJoystickHandler.cs:228]: Given that just a few lines above, the `axisValue` is set to be the absolute value of the raw axis value, unless cosmic rays decide to intervene it should be impossible to have a negative deadzone. That said, I'll switch to a clamp for clarity.

Comment (bdach) [reply osu.Framework/Input/Handlers/Joystick/OsuTKJoystickHandler.cs:228]: Fair enough, wouldn't bother with clamping then.

**Comment by bdach** [comment]
Seems like this PR is doing what it intends to.

As a vaguely related discussion, not sure whether using initial axis value readouts to seed the default deadzones is a good choice (consider launching a game with the stick/triggers held to their extreme values - unless the game specifically implements deadzone adjustment/calibration, it will be stuck with mostly non-functional input), but that is not really in scope here and a separate thing entirely.

**Comment by frenzibyte** [comment]
Yeah, and there was an [old issue](https://github.com/ppy/osu-framework/issues/2854) opened about triggers in the Xbox 360 Controller not being handled properly, which turned out to be because of this exact logic.

I believe [the ability to configure joystick axis deadzones](https://github.com/ppy/osu-framework/issues/3639) should leverage this behaviour to at least always have usable joysticks.

**Comment by peppy** [comment]
Tested on the only controller I have on hand and seems to work as expected (doesn't break). Let's get this in.

**Pull request merged**